var/spack/repos/builtin/packages/openblas/package.py

```diff
@@ -110,6 +114,10 @@ def make_defs(self):
                 'TARGET=PILEDRIVER',
                 'TARGET=ARMV8'
             ]
+        if self.spec.variants['cpu_target'].value:
```

Comment:
We also set targets directly above this line. Will that be a problem?

Reply:
> I'm unsure what spack.architecture.sys_type() is.

That gives you a 3-part architecture string with platform, operating system, and target. For example:

`linux-rhel6-x86_64`

You can access the target directly using `architecture.platform().target(id)`. `id` can be an arbitrary string, or there are special values you can use: `frontend`, `default_target`, or `backend`. Ideally this function could be used to automatically populate `TARGET=`. I imagine there could be issues where even if the auto-discovered target is "close", it doesn't exactly match what this build system is expecting. The preceding lines (e.g. `TARGET=PILEDRIVER`) appear to be an example of that.

Are you using this new variant to set several different values of `TARGET`? In particular are you doing so for a fixed value of `architecture.sys_type()`? If you are setting one value of `TARGET`, perhaps you could check what `architecture.sys_type()` looks like in your case and follow the same strategy as for 'aarch64'.

@becker33 do you have any opinions on this?

Reply:
If I remember correctly it turned out that doing 
```cmake TARGET=PILEDRIVER TARGET=ARMV8```
would leave ```TARGET=ARMV8``` (although I'm not sure whether it was the first or the last occurrence that ended up being used). We needed it to explicitly set the target architecture as the frontends have older CPUs than the cluster nodes, making the execution of ```matplotlib``` on the frontend failing with an "invalid operation" (or something similar) error.

I would suspect that using ```spack```'s auto detection would end up (ideally) finding the same results that ```openblas``` is already detecting.

Reply:
> We needed it to explicitly set the target architecture as the frontends have older CPUs than the cluster nodes, making the execution of matplotlib on the frontend failing with an "invalid operation" (or something similar) error.

In your case were you building with something like `spack spec <package> os=fe target=fe`? In that case I assume you could use the same strategy as was done for `aarch64`

Reply:
No we were just building as ```spack install <some-package-depending-on-openblas>```. Our current solution is in ```packages.yaml```
```
  openblas:
    variants: ['cpu_target=NEHALEM']
```

Trying your syntax leads to the following error:
```
$ spack spec openblas target=NEHALEM
Input spec
--------------------------------
openblas arch=linux-None-NEHALEM 

Normalized
--------------------------------
openblas arch=linux-None-NEHALEM 

Concretized
--------------------------------
==> Error: No compilers found for operating system debian8 and target NEHALEM.
If previous installations have succeeded, the operating system may have been updated.
Compilers are defined for the following operating systems and targets:
        debian8-x86_64
Run 'spack compiler find' to add compilers.
```
I'd rather not have to add all compilers again just to be able to specify the target architecture for one package

Reply:
Regarding:

> `spack spec openblas target=NEHALEM`

Apologies but I should have clarified: `fe` is a shorthand for `frontend` which has a special meaning: use the frontend target or operating system. You should be able to use the exact string:

`spack spec obenblas os=fe target=fe`

(you may have to set "os" and "target")

Reply:
This has no visible effect..
```
$ spack spec -L openblas
Input spec
--------------------------------
iqz7m24eghgqz7jqymlr5evemgqzarif  openblas

Normalized
--------------------------------
iqz7m24eghgqz7jqymlr5evemgqzarif  openblas

Concretized
--------------------------------
4jl2pvp4im2skyqjabkfdpxhm56evyyn  openblas@0.2.20%gcc@4.8~openmp+pic+shared arch=linux-debian8-x86_64 

$ spack spec -L openblas os=fe target=fe
Input spec
--------------------------------
7tdz63dul4ta3ftrrigikvalqzojyfme  openblas arch=linux-debian8-x86_64 

Normalized
--------------------------------
7tdz63dul4ta3ftrrigikvalqzojyfme  openblas arch=linux-debian8-x86_64 

Concretized
--------------------------------
4jl2pvp4im2skyqjabkfdpxhm56evyyn  openblas@0.2.20%gcc@4.8~openmp+pic+shared arch=linux-debian8-x86_64 
```
I tried to find the code part where the ```os=``` and ```target=``` is handled, could you give me a pointer?

Reply:
I looked into this and I think our architecture detection is not fine-grained enough to distinguish `NEHALEM` and furthermore it wouldn't make sense to handle this in the same way as `aarch64`.

So my only request here given that you mention

> This would mean that only 'TARGET'=ARMV8' has any effect.

Is to take `if 'aarch64' in spack.architecture.sys_type():` and place it after your addition and make it an `elif` (so if `cpu_target` is set, do that, otherwise if `aarch64` is in `sys_type`...).

> I tried to find the code part where the os= and target= is handled, could you give me a pointer?

A few places:

* The `fe`/`be` shorthand is established in `architecture.py` in the `Platform` class
* The command line parsing is in `spec.py` (handled by `SpecParser.do_parse`)
* `concretize.py` deals with assigning these to specs where they aren't explicitly set

Reply:
Done